collectors/proc.plugin/sys_class_infiniband.c

```diff
@@ -0,0 +1,721 @@
+							strcat(HW->file_##NAME, "/"#NAME);
+
+						// VENDOR-MLX: Mellanox
+						if (strncmp(dev_dent->d_name, "mlx", 3) == 0) {
+							// Allocate the vendor specific struct
+							p->hwcounters_mlx = callocz(1, sizeof(struct ibporthw_mlx));
+
+							FOREACH_HWCOUNTER_MLX(GEN_DO_HWCOUNTER_NAME, p, p->hwcounters_mlx)
+
+							// Set the function pointer for hwcounter parsing
+							p->hwcounters_parse  = &infiniband_hwcounters_parse_mlx;
+							p->hwcounters_dorrd = &infiniband_hwcounters_dorrd_mlx;
+						}
+
+						// VENDOR: Unknown
+						else {
+							p->do_hwpackets = CONFIG_BOOLEAN_NO;
+							p->do_hwerrors  = CONFIG_BOOLEAN_NO;
+						}
+					}
+				}
+
+				// Check / Update the link speed & width frm "rate" file
+				// Sample output: "100 Gb/sec (4X EDR)"
+				snprintfz(buffer, FILENAME_MAX, "%s/%s/%s", ports_dirname, port_dent->d_name, "rate");
+				char buffer_rate[65];
+				if (read_file(buffer, buffer_rate, 64)) {
+					error("Unable to read '%s'", buffer);
+					p->width = 1;
+				}
+				else {
+					char *buffer_width = strstr(buffer_rate, "(");
+					buffer_width++;
+					// str2ull will stop on first non-decimal value
+					p->speed = str2ull(buffer_rate);
+					p->width = str2ull(buffer_width);
+				}
+
+				info("Infiniband card %s port %s at speed %llu width %llu", dev_dent->d_name, port_dent->d_name, p->speed, p->width);
+
+			}
+			closedir(ports_dir);
+		}
+		closedir(devices_dir);
+
+		initialized = 1;
+	}
+
+	// Update all ports values
+	struct ibport *port;
+	for (port = ibport_root; port ; port = port->next) {
+
+		if (!port->enabled)
+			continue;
+
+		//
+		// Read values from system to struct
+		//
+
+		//  counter from file and place it in ibport struct
+		#define GEN_DO_COUNTER_READ(NAME,GRP,DESC,DIR,PORT, ...) \
+			if (PORT->file_##NAME) { \
+				if (read_single_number_file(PORT->file_##NAME, (unsigned long long *) &PORT->NAME)) { \
+					error("cannot read iface '%s' counter '"#NAME"'", PORT->name); \
+					PORT->file_##NAME = NULL; \
+				} \
+			}
+
+		// Update charts
+		if (port->do_bytes != CONFIG_BOOLEAN_NO) {
+
+			// Read values from sysfs
+			FOREACH_COUNTER_BYTES(GEN_DO_COUNTER_READ, port)
+
+			// First creation of RRD Set (charts)
+			if(unlikely(!port->st_bytes)) {
+				port->st_bytes = rrdset_create_localhost(
+					"Infiniband"
+					, port->chart_id_bytes
+					, NULL
+					, port->chart_family
+					, "ib.bytes"
+					, "Bandwidth usage"
+					, "kilobits/s"
+					, PLUGIN_PROC_NAME
+					, PLUGIN_PROC_MODULE_INFINIBAND_NAME
+					, port->priority + 1
+					, update_every
+					, RRDSET_TYPE_AREA
+				);
+				// Create Dimensions
+				rrdset_flag_set(port->st_bytes, RRDSET_FLAG_DETAIL);
+				// On this chart, we want to have a KB/s so the dashboard will autoscale it
+				// The reported values are also per-lane, so we must multiply it by the width
+				FOREACH_COUNTER_BYTES(GEN_RRD_DIM_ADD_CUSTOM, port, 8 * port->width, 1024, RRD_ALGORITHM_INCREMENTAL)
```

Comment:
I just have a double about the multiplier `port->width` is working, as I didn't see any higher than 24Gb/s on my test. 
I don't know if I didn't get the MPI benchmark with the correct options to saturate the port, or if the multiplier is not taken into account. From the "info" log line, it's correctly reported as 4.